ndimage: implement higher order spline interpolation

grlee77 opened the pull request:
closes #1523

blocked by #4400 and #4401

### Overview

This PR implements spline interpolation orders 2-5. Spline interpolation was substantially improved in SciPy 1.6, with all boundary modes now being handled correctly. Only the `mode='mirror'` cases can be compared tested against older SciPy.

**To decide**: How do we want to handle the order argument to these functions. Should we change the default back from None to 3 so that it matches SciPy? For now, I have left it as it was in CuPy 8.0.

### Performance

I will post some benchmark results when I get a chance. This is still fast relative to SciPy, but the acceleration is worse than that seen for orders 0 and 1. I would like to see if there is a way to refactor this a bit to reduce register usage, as I suspect that is part of the reason for the performance difference.

takagi commented:
Thanks! As for the default value of `order`, yes, it should be changed to align with SciPy. I put `no-compat` label.

takagi commented:
pfnCI, test this please.

grlee77 commented:
I have changed the default order to 3 as in SciPy and updated the docstrings accordingly.

Please try running the tests again (I fixed a missing argument that was due to a rebasing mistake). All interpolation tests are passing locally.

takagi commented:
pfnCI, test this please.

takagi commented:
Thanks! Waiting for CI pass and will be merged soon.
